libs/fairness/src/lib/util/calculateFairnessMetric.test.ts

```diff
@@ -0,0 +1,493 @@
+  });
+  it("should calculate ratio with 3+ groups correctly", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.25, upper: 0.4 },
+        { lower: 0.6, upper: 0.75 },
+        { lower: 0.45, upper: 0.55 }
+      ],
+      bins: [0.35, 0.7, 0.5],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.33), upper: closeTo(0.67) },
+      overall: closeTo(0.5)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate ratio with 3+ groups correctly", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.45, upper: 0.55 },
+        { lower: 0.25, upper: 0.4 },
+        { lower: 0.6, upper: 0.75 }
+      ],
+      bins: [0.5, 0.35, 0.7],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.33), upper: closeTo(0.67) },
+      overall: closeTo(0.5)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate ratio with 3+ groups correctly (with overlap)", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.45, upper: 0.65 },
+        { lower: 0.25, upper: 0.4 },
+        { lower: 0.6, upper: 0.75 }
+      ],
+      bins: [0.5, 0.35, 0.7],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.33), upper: closeTo(0.67) },
+      overall: closeTo(0.5)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should calculate ratio with 3+ groups correctly (with all overlap)", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.45, upper: 0.7 },
+        { lower: 0.25, upper: 0.55 },
+        { lower: 0.45, upper: 0.75 }
+      ],
+      bins: [0.5, 0.35, 0.7],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      bounds: { lower: closeTo(0.33), upper: closeTo(1) },
+      overall: closeTo(0.5)
+    };
+    expect(result).toMatchObject(expectedResult);
+  });
+  it("should handle ratio case with large binBound that provides the largest and smallest bounds", () => {
+    const mockValue = {
+      binBounds: [
+        { lower: 0.1, upper: 0.3 },
+        { lower: 0.7, upper: 0.9 },
+        { lower: 0.08, upper: 0.95 }
+      ],
+      bins: [0.2, 0.8, 0.5],
+      global: 0.5
+    };
+    const result = calculateFairnessMetric(mockValue, FairnessModes.Ratio);
+    const expectedResult = {
+      // from the 1st and 3rd point: the lower bound = 0.1 / 0.95
+      // from the 1st and 2nd point: the upper bound = 0.3 / 0.7
+      // the smallest ratio between nominal bins: overall = 0.2 / 0.8
+      bounds: { lower: closeTo(0.09), upper: closeTo(0.43) },
```

Comment:
This comment is confusing me - isn't the critical one 0.08/0.9 = 0.09?

Reply:
good catch